extensions/rest-client/runtime/src/main/java/io/quarkus/restclient/runtime/RestClientRecorder.java

```diff
@@ -59,25 +59,36 @@ public InjectorFactory getInjectorFactory() {
             }
         };
 
-        if (useBuiltIn) {
-            RegisterBuiltin.register(clientProviderFactory);
-            registerProviders(clientProviderFactory, contributedProviders, false);
+        registerProviders(clientProviderFactory, useBuiltIn, providersToRegister, contributedProviders);
+
+        if (ResteasyProviderFactory.peekInstance() != null) {
```

Comment:
Does the execution order play a role here? @jaikiran reported that sometimes the server one was initialized first and sometimes the client one.

I think we should have a build item ensuring that the order of operations is fully determined.

Reply:
To my understanding, the only time there's a problem is when the server one is created first and the REST Client one clobbers it.

I don't think the reverse situation poses any problems.

In addition, there are many times where only the server or only REST Client is present, so I'm not sure how a BuildItem helps in that situation

Reply:
Another point to note is that the provider factory for server or client side are actually different instances of the same class, where the "server" instance has a no-op for all client operations, and the "client" instance the reverse.

Reply:
We could have a build item in one of the artifacts common to the two, the server would produce it and the client optionally consume it.

This way, if the server is present, we would be sure the client code would be executed after the server code.

Obviously, we can also do the opposite if it makes more sense.

Reply:
Another option, which will take some time, is to add the functionality we have for `RestClientBuilderImpl` to `ResteasyClientBuilderImpl` that can use a static method to set the provider directly into the instance.

Then ordering isn't an issue. I've raised https://issues.redhat.com/browse/RESTEASY-2684 to discuss this option with the RESTEasy team

Reply:
While fixing things in RESTEasy would be nice, I think we need a reasonable fix for 1.8.

Let's wait for @jaikiran 's opinion on this.

Reply:
Yes, we need to be sure it fixes the random ordering.

Reply:
So this change still leaves the chance for the ordering issue to occur. The order I'm talking about is as follows:

1. There are 2 separate extensions in play here - `quarkus-rest-client` and `quarkus-resteasy`.

2. Each of them have multiple build steps (which is fine). Both of these extensions, in one of their build steps "setup" a `@Recorder` of their own to call some method which ultimately leads to a call to `ResteasyProviderFactory.setInstance` method. In case of the `quarkus-rest-client` extension, the recorder in question is the `RestClientRecorder` and for `quarkus-resteasy` extension it is the `ResteasyStandaloneRecorder`. So far all is fine.

3. In Quarkus, these recorders get invoked upon application initialization. It so happens that the order in which these are recorders are invoked isn't deterministic (from what I saw). So a few times the `quarkus-resteasy` recorder might get invoked first and in some cases the `quarkus-rest-client` recorder might get invoked first. Ultimately, whichever recorder gets invoked last "wins" the call to `ResteasyProviderFactory.setInstance` by overwriting the instance.

This issue still exists with this change because if the `quarkus-rest-client` recorder gets invoked first it will peek and see that the `instance` is null (assuming no other extension within the Quarkus ecosystem ended up setting it). Since it is null, it will go ahead and set the instance to `clientProviderFactory`. Now next, after a while, the `quarkus-resteasy` recorder gets invoked its recorder ultimately invokes certain method which leads to[1] which will overwrite this instance, which was set by the `rest-client` recorder.
So the ordering issue remains. Having said that, to answer this:

> To my understanding, the only time there's a problem is when the server one is created first and the REST Client one clobbers it.
> I don't think the reverse situation poses any problems

I'm not too sure, because outside of the current issue, I don't have much idea about how resteasy behaves internally with these provider factory instances.

@gsmet, does ordering (via build item dependencies) of build steps (across extensions) also order their recorder execution?

[1] https://github.com/resteasy/Resteasy/blob/4.5.6.Final/resteasy-core/src/main/java/org/jboss/resteasy/core/ResteasyDeploymentImpl.java#L373

Reply:
@jaikiran the recorder executions will be executed in the order they are recorded. So as soon as the build steps are properly ordered, it's going to work.

So, in conclusion, please make sure the build steps playing with `ResteasyProviderFactory` are executed in order via build items dependencies.

Reply:
@jaikiran I know what you've outlined. However, the server provider factory overriding the rest client created one is not actually an issue as far as I've seen.

In such a situation, the providers rest client needs are registered by the server provider factory, and rest client works as expected.

The only problem situation is when rest client overrides the server provider factory, which this PR solves

Reply:
> In such a situation, the providers rest client needs are registered by the server provider factory, and rest client works as expected.

Hello @kenfinnigan, that sounds fine to me then.

Reply:
Again, I want to be sure we properly control the execution order.

@kenfinnigan I can do the work and push an additional commit to your PR if you don't think it's useful.

Reply:
I don't see it as necessary or useful

Reply:
To reiterate, with this change it's not necessary to control the execution order. And I'm looking to properly fix it in RESTEasy so this wouldn't be needed long term either.

I don't see the need to further complicate the code for a change that isn't permanent

Reply:
> And I'm looking to properly fix it in RESTEasy so this wouldn't be needed long term either.

Yeah but that won't be for 1.8 and we need a fix now.

And I really think having a deterministic execution order for things that interact with each other is important.

If just for the sake of not spending hours debugging weird issues.

I think adding the build item should be very easy. I'll take a look tomorrow. Either it's easy and don't add boilerplate or I will give up.

Reply:
I know it won't be done for 1.8, but why add more code that needs to be removed when it's there in 1.9?

When the change is done in RESTEasy, there is no longer a need for a build item